File: features/miso/loop/dictate/as-posts/capture/video/flip/while-recording/segments/segments.js
// a flip mid-take closes the clip and opens another on the other camera, and
// the mini joins the pieces into one file afterwards.
//
// /while-recording drew the camera onto a canvas and recorded THAT, so the
// recorder never saw a track change. It works, and it costs a copy of every
// frame for the whole take and resamples the camera onto a second clock — when
// the phone is busy the animation tick slips and the same frame is captured
// twice, which reads as a stutter. Ash asked for the other road (#p139): record
// the camera natively, cut a new recording at the flip, join afterwards.
//
// So this node stands the canvas down through the seam its parent opened, and
// takes the flip itself.
const feature_Segments = {
  // a container shorter than this is not worth making: MediaRecorder will
  // write one, but a fraction of a second of mp4 is a header and little else,
  // and ffmpeg's concat demuxer is entitled to dislike it. Under it, the
  // segment is dropped and the take starts clean on the new camera.
  MIN_MS: 300,

  marks: [],        // the part index at which each container starts: [0, 7, …]
  segStart: 0,      // when the current segment's recorder was started
  swapping: false,

  // ---- the camera, natively ------------------------------------------------
  // the recorder is handed /video's own stream, which is what it recorded
  // before /while-recording existed. Its canvas, its draw loop and its source
  // element are never made, and `drawing` stays false so its own flip watcher
  // does nothing — this node's takes its place.
  begin() {
    this.marks = [0];
    this.segStart = Date.now();
    // a new take starts clean whatever the last one left behind. A cut that
    // never finished — the camera prompt was up and `getUserMedia` waited
    // behind it while the take was stopped — would otherwise leave this set
    // and every flip of every later take would be ignored. Rig-found,
    // 2026-09-04.
    this.swapping = false;
    // and the camera this take opened on, so the first paint is not read as a
    // flip: the watcher compares against it.
    feature_WhileRecording.facing = feature_WhileRecording.wanted();
    return feature_Video.media;
  },

  // ---- the flip ------------------------------------------------------------
  // noticed at the paint, exactly as /while-recording noticed it: the answer
  // the whole chain gives for the camera, whoever composed it.
  onApply() {
    if (this.swapping) return;
    if (typeof feature_Video === 'undefined' || !feature_Video.recorder) return;
    if (feature_Video.recorder.state !== 'recording') return;
    const want = feature_WhileRecording.wanted();
    if (!want || want === feature_WhileRecording.facing) return;
    this.cut(want);
  },

  // wait for a recorder to have handed over its last piece. `onstop` fires
  // after the final `dataavailable`, so this resolves with every byte of the
  // segment already in `chunks` and already posted by /streams.
  settled(rec) {
    return new Promise((done) => {
      let over = false;
      const end = () => { if (!over) { over = true; done(); } };
      rec.onstop = end;
      rec.onerror = end;
      setTimeout(end, 4000);          // a recorder that never stops is not a reason to hang
      try { rec.stop(); } catch (e) { end(); }
    });
  },

  async cut(facing) {
    const v = feature_Video;
    this.swapping = true;
    let next = null;
    try {
      // the new camera FIRST: if it refuses, the take carries on untouched and
      // nothing has been closed.
      next = await navigator.mediaDevices.getUserMedia({
        video: { facingMode: facing }, audio: false });
    } catch (e) {
      feature_WhileRecording.facing = feature_WhileRecording.wanted();
      this.swapping = false;
      return;
    }
    // the camera can take a while to answer — a permission sheet stands in
    // front of it the first time — and the take may be over by the time it
    // does. Nothing is closed in that case and the camera just opened is let
    // go again.
    if (!v.recorder || v.recorder.state !== 'recording') {
      next.getTracks().forEach((t) => t.stop());
      this.swapping = false;
      return;
    }
    const short = Date.now() - this.segStart < this.MIN_MS;
    const rec = v.recorder;
    await this.settled(rec);

    // a segment too short to be a container is thrown away and the take starts
    // clean on the new camera. Nothing has been posted yet — the timeslice is
    // two seconds and this is under a third of one — so there is nothing on
    // the exchange to contradict.
    if (short) {
      v.chunks = [];
      this.marks = [0];
      if (typeof feature_Streams !== 'undefined') { feature_Streams.count = 0; }
    } else {
      this.marks.push(v.chunks.length);
    }

    // the microphone never moves: it is the same microphone whichever way the
    // camera points, and a second one would be a second voice in the room.
    const audio = v.media ? v.media.getAudioTracks() : [];
    const tracks = [next.getVideoTracks()[0]];
    if (audio.length) tracks.push(audio[0]);
    const stream = new MediaStream(tracks);

    // release the camera we came off — /video's own stream keeps its
    // microphone, which the new recorder is still holding.
    const old = feature_WhileRecording.cam;
    if (old === v.media) {
      old.getVideoTracks().forEach((t) => t.stop());
    } else if (old) {
      old.getTracks().forEach((t) => t.stop());
    }
    feature_WhileRecording.cam = next;
    feature_WhileRecording.facing = facing;
    feature_WhileRecording.extra.push(next);   // its teardown releases them

    try {
      v.recorder = new MediaRecorder(stream, v.opts());
      // /video's own handler, rebuilt: the part number keeps counting across
      // segments, so /streams' `parts` and the server's join loop see one
      // unbroken run of pieces whatever happened in the middle.
      v.recorder.ondataavailable = (e) => {
        if (!e.data.size) return;
        v.chunks.push(e.data);
        v.onChunk(e.data, v.chunks.length - 1);
      };
      v.recorder.onstop = () => v.save();
      v.recorder.start(v.timeslice());
      this.segStart = Date.now();
    } catch (e) {
      // no recorder: the take is over rather than silently dead
      feature_Loop.send({ type: 'click', ev: 'vid_stop' });
    }

    // the viewfinder follows at once — a flip you cannot see is a flip that
    // lies about which way the phone is pointing
    if (v.view) {
      v.view.srcObject = stream;
      const p = v.view.play(); if (p && p.catch) p.catch(() => {});
    }
    this.swapping = false;
  },

  // ---- what the mini is told -----------------------------------------------
  // the marks ride on the clip's own metadata, beside /streams' `parts`, so
  // they reach the exchange on the RecShared that announces the recording and
  // land on its index with everything else. One segment is no marks at all:
  // the server then joins the pieces by concatenation, exactly as today.
  metaFor(meta) {
    if (this.marks.length < 2) return meta;
    return Object.assign({}, meta, { segs: this.marks.slice() });
  },

  // ---- what this phone plays -----------------------------------------------
  // the local copy is every piece run together, which is several containers in
  // one file: a player reads the first and stops. That is the first segment,
  // and it is honest — but it is not the take. Once the mini has joined the
  // pieces, `blob/<id>` is the whole thing, so the next play of a multi-segment
  // clip asks for it and keeps it. If the exchange has not joined it yet (or
  // is not reachable) the first segment is what plays, and the next play tries
  // again. Nothing is ever deleted: a take always has something to show.
  installFetch() {
    if (typeof feature_Dictate === 'undefined' || feature_Dictate.fm_segments) return false;
    feature_Dictate.fm_segments = true;
    const orig = feature_Dictate.getBlob.bind(feature_Dictate);
    feature_Dictate.getBlob = async (id) => {
      const local = await orig(id);
      if (!local) return local;
      let meta = null;
      try { meta = await feature_Dictate.getBlob('meta:' + id); } catch (e) {}
      if (!meta || !meta.segs || meta.segs.length < 2 || meta.joined) return local;
      try {
        const r = await fetch('blob/' + id);
        if (!r.ok) return local;
        const whole = await r.blob();
        // any file at all under this id is the joined one: the exchange only
        // writes `blob/<id>` once ffmpeg has succeeded, and 404s until then.
        // Its SIZE is not the test — a re-mux has one header where the pieces
        // had four, so the joined file is often the smaller of the two.
        if (!whole || !whole.size) return local;
        await feature_Dictate.put(id, whole);
        await feature_Dictate.put('meta:' + id, Object.assign({}, meta, { joined: true }));
        return whole;
      } catch (e) { return local; }
    };
    return true;
  },

  install() {
    if (typeof feature_WhileRecording === 'undefined' || typeof feature_Video === 'undefined') {
      return false;
    }
    // /streams REPLACES metaFor rather than wrapping it, and installs off an
    // interval of its own — so this one waits for it before wrapping, or there
    // is a race in which the marks are quietly dropped. With /streams not
    // composed there is nothing to wait for.
    if (typeof feature_Streams !== 'undefined' && !feature_Video.fm_streams) {
      return false;
    }
    if (feature_WhileRecording.fm_segments) return true;
    feature_WhileRecording.fm_segments = true;
    const self = this;
    // the canvas road stands down: the parent's own seam, answered with the
    // camera's stream and nothing built.
    feature_WhileRecording.begin = function () { return self.begin(); };
    feature_WhileRecording.onApply = function () { self.onApply(); };
    const fm_segMeta = feature_Video.metaFor.bind(feature_Video);
    feature_Video.metaFor = function (meta) { return self.metaFor(fm_segMeta(meta)); };
    self.installFetch();
    return true;
  },
};

{
  let fm_segTries = 0;
  const fm_segInit = setInterval(() => {
    fm_segTries = fm_segTries + 1;
    if (feature_Segments.install() || fm_segTries > 100) clearInterval(fm_segInit);
  }, 100);
}

File: features/miso/loop/dictate/as-posts/capture/video/flip/while-recording/segments/segments.rs
struct feature_Segments;
impl feature_Segments {
    // ---- the marks reach the parts directory before the join --------------
    // /streams joins on the RecShared that announces a recording, and it joins
    // BEFORE the rest of the chain runs — which is where /mirror writes the
    // index. So the marks cannot be read off the index at that moment: they
    // are not there yet. This link is outermost, so it writes them beside the
    // pieces first and the join finds them however it is reached — by the
    // announcement, or by a piece arriving late afterwards.
    fn handle_msg(msg: String) -> String {
        let m: serde_json::Value = serde_json::from_str(&msg)
            .unwrap_or(serde_json::Value::Null);
        if m["type"].as_str().unwrap_or("") != "RecShared" {
            return existing.handle_msg(msg);
        }
        let id = m["data"]["id"].as_str().unwrap_or("").to_string();
        let segs = m["data"]["segs"].clone();
        if streams_id_ok(&id) && segs.is_array() {
            let from = m["_from"].as_str().unwrap_or("").to_string();
            let world = if from.is_empty() { "_local".to_string() } else { from };
            let dir = streams_parts_dir(world, id);
            let _ = std::fs::create_dir_all(dir.clone());
            let _ = std::fs::write(format!("{}/segs.json", dir), segs.to_string());
        }
        existing.handle_msg(msg)
    }

    // ---- the join ----------------------------------------------------------
    // one segment is one container cut at byte boundaries, and /streams' own
    // concatenation is exactly right for it. Two or more are two or more
    // CONTAINERS, and gluing those end to end gives a file whose first header
    // describes only the first take — a player reads that and stops. So the
    // pieces of each segment are run together into a file of its own and
    // ffmpeg is asked to concatenate those.
    //
    // `-c copy`, because every segment came off the same phone through the
    // same MediaRecorder: iOS hands back `video/mp4; codecs=avc1.42000a,
    // mp4a.40.2` for all of them, so the streams match and there is nothing to
    // re-encode. A mismatch would need a transcode; it is named in the spec and
    // not built, because a phone that changed codec mid-take is not a thing
    // that happens and building for it blind would be worse than saying so.
    fn streams_join(world: String, id: String, want: u64) -> bool {
        let dir = streams_parts_dir(world.clone(), id.clone());
        let marks = segments_marks(dir.clone(), want);
        if marks.len() < 2 {
            return existing.streams_join(world, id, want);
        }
        let file = format!("{}/{}/{}", streams_root(), world, id);
        if std::path::Path::new(&file).exists() {
            let _ = std::fs::remove_dir_all(dir);
            return false;                 // already whole; nothing new happened
        }
        // every piece must be here before anything is written: a segment file
        // built from half its pieces is a file that looks finished and is not
        let mut n: u64 = 0;
        while n < want {
            if !std::path::Path::new(&format!("{}/{}", dir, n)).exists() {
                return false;             // a piece is still on its way
            }
            n = n + 1;
        }
        let ff = segments_ffmpeg();
        if ff.is_empty() {
            println!("segments: no ffmpeg on this machine — {} left in pieces", id);
            return false;
        }
        // through MPEG-TS, and this is the whole reason the join is not one
        // command. The concat DEMUXER offsets each input by what the one before
        // it says its duration is — and a MediaRecorder mp4 is written
        // incrementally, so what its header says is not always what it holds.
        // Joined that way, three takes came out right and the fourth claimed
        // 991 seconds for ten seconds of video: 434 frames with packet stamps
        // running to 947 s, the audio beside it a sane 10.1 (rig-found,
        // 2026-09-04). TS carries no global duration to be wrong about, so
        // each segment is copied into it and the pieces are concatenated as
        // byte streams. Still no re-encode: two copy passes and a bitstream
        // filter each way, which is the standard road for h264+aac.
        let mut chain = String::new();
        let mut k = 0usize;
        while k < marks.len() {
            let from = marks[k];
            let to = if k + 1 < marks.len() { marks[k + 1] } else { want };
            let seg = format!("{}/seg-{}.mp4", dir, k);
            let ts = format!("{}/seg-{}.ts", dir, k);
            if !segments_write_one(dir.clone(), from, to, seg.clone()) {
                return false;
            }
            if !segments_ff(ff.clone(), vec![
                "-i".to_string(), seg.clone(),
                "-c".to_string(), "copy".to_string(),
                "-bsf:v".to_string(), "h264_mp4toannexb".to_string(),
                "-f".to_string(), "mpegts".to_string(),
                ts.clone()], id.clone()) {
                return false;
            }
            if !chain.is_empty() {
                chain.push('|');
            }
            chain.push_str(&ts);
            k = k + 1;
        }
        let _ = std::fs::create_dir_all(format!("{}/{}", streams_root(), world));
        let out = format!("{}/joined.mp4", dir);
        if !segments_ff(ff, vec![
            "-i".to_string(), format!("concat:{}", chain),
            "-c".to_string(), "copy".to_string(),
            "-bsf:a".to_string(), "aac_adtstoasc".to_string(),
            "-movflags".to_string(), "+faststart".to_string(),
            out.clone()], id.clone()) {
            return false;
        }
        if std::fs::rename(&out, &file).is_err() {
            // a rename across devices can refuse; the copy is the fallback
            match std::fs::read(&out) {
                Ok(bytes) => {
                    if std::fs::write(&file, &bytes).is_err() {
                        println!("segments: could not write the joined {}", id);
                        return false;
                    }
                }
                Err(_) => return false,
            }
        }
        let _ = std::fs::remove_dir_all(dir);
        println!("segments: {} joined from {} segments over {} pieces",
                 id, marks.len(), want);
        true
    }

    // one ffmpeg run, with `-y` and quiet logging in front of whatever it is
    // asked to do. A refusal is said in the log and answered `false`, so the
    // clip stays in pieces rather than being written half-joined.
    fn segments_ff(ff: String, args: Vec<String>, id: String) -> bool {
        let mut cmd = std::process::Command::new(ff);
        cmd.arg("-y").arg("-hide_banner").arg("-loglevel").arg("error");
        for a in args.iter() {
            cmd.arg(a);
        }
        match cmd.output() {
            Ok(o) => {
                if !o.status.success() {
                    println!("segments: ffmpeg refused {} — {}", id,
                             String::from_utf8_lossy(&o.stderr).replace('\n', " "));
                }
                o.status.success()
            }
            Err(e) => {
                println!("segments: could not run ffmpeg for {} — {}", id, e);
                false
            }
        }
    }

    // the pieces of one segment, run together as they were cut. Within a
    // segment this is /streams' own concatenation and is right for the same
    // reason: MediaRecorder's timeslices are one container cut at byte
    // boundaries.
    fn segments_write_one(dir: String, from: u64, to: u64, out: String) -> bool {
        let mut all: Vec<u8> = Vec::new();
        let mut n = from;
        while n < to {
            match std::fs::read(format!("{}/{}", dir, n)) {
                Ok(mut bytes) => all.append(&mut bytes),
                Err(_) => return false,
            }
            n = n + 1;
        }
        if all.is_empty() {
            return false;
        }
        std::fs::write(&out, &all).is_ok()
    }

    // the marks the phone sent, read back off the sidecar: the part index each
    // container starts at. Anything that is not a run of ascending indices
    // inside the piece count is not trusted and the clip is joined the old way
    // — a bad mark would cut a segment in the wrong place, and a file that
    // plays its first take is a better failure than one that plays nothing.
    fn segments_marks(dir: String, want: u64) -> Vec<u64> {
        let raw = std::fs::read_to_string(format!("{}/segs.json", dir))
            .unwrap_or_default();
        let v: serde_json::Value = serde_json::from_str(&raw)
            .unwrap_or(serde_json::Value::Null);
        let empty: Vec<serde_json::Value> = Vec::new();
        let arr = v.as_array().unwrap_or(&empty);
        let mut out: Vec<u64> = Vec::new();
        let mut last: i64 = -1;
        for m in arr.iter() {
            let n = match m.as_u64() {
                Some(n) => n,
                None => return Vec::new(),
            };
            if n >= want || (n as i64) <= last {
                return Vec::new();
            }
            last = n as i64;
            out.push(n);
        }
        if out.is_empty() || out[0] != 0 {
            return Vec::new();
        }
        out
    }

    // where ffmpeg is. The server is started by launchd with a PATH that has
    // no brew in it (deploy.md), so the usual places are tried by name before
    // the PATH is trusted at all.
    fn segments_ffmpeg() -> String {
        let mut tries: Vec<String> = Vec::new();
        if let Ok(set) = std::env::var("MISO_FFMPEG") {
            tries.push(set);
        }
        tries.push("/opt/homebrew/bin/ffmpeg".to_string());
        tries.push("/usr/local/bin/ffmpeg".to_string());
        tries.push("/usr/bin/ffmpeg".to_string());
        for t in tries.iter() {
            if std::path::Path::new(t).exists() {
                return t.clone();
            }
        }
        match std::process::Command::new("which").arg("ffmpeg").output() {
            Ok(o) => {
                let p = String::from_utf8_lossy(&o.stdout).trim().to_string();
                if !p.is_empty() && std::path::Path::new(&p).exists() {
                    return p;
                }
                String::new()
            }
            Err(_) => String::new(),
        }
    }
}
